sunpy/io/special/genx.py

```diff
@@ -0,0 +1,140 @@
+    }
+    for key in subskeleton:
+        if isinstance(subskeleton[key], OrderedDict):
+            struct_to_data(xdrdata, subskeleton[key])
+        elif isinstance(subskeleton[key], np.ndarray):
+            testlist = list()
+            struct_shape = subskeleton[key].shape
+            for elem in subskeleton[key].flatten():
+                elem2 = copy.deepcopy(elem)
+                struct_to_data(xdrdata, elem2)
+                testlist.append(elem2)
+            subskeleton[key] = np.array(testlist).reshape(struct_shape)
+        else:
+            sswsize = subskeleton[key]
+            sswtype = sswsize[-2]
+            if sswsize[0] == 0:
+                subskeleton[key] = types_dict[sswtype][0]()
+            else:
+                subskeleton[key] = np.array(xdrdata.unpack_farray(sswsize[-1], types_dict[sswtype][0]),
+                                            dtype=types_dict[sswtype][1]).reshape(sswsize[1:-2][::-1])
+
+def read_genx(filename):
+    """
+    solarsoft genx file reader
+
+    Parameters
+    ----------
+    filename : `str`
+        The genx file to be read
+
+    Returns
+    -------
+    output : `OrderedDict`
+        A dictionary with possibly nested dictionaries with the data in
+        the genx file.
+
+    Notes
+    -----
+    The reader aims to maintain the shape and type of the arrays, but take care with the
+    difference in indexing between Python and IDL.
+    Regarding the type notice that single numbers are converted to python precision, therefore
+    a single integer is converted from 16 to 32/64 bits, and a float from 32 to 64.
+    """
+    with open(filename, mode='rb') as xdrfile:
+        xdrdata = ssw_xdrlib(xdrfile.read())
+
+    # HEADER information
+    version, xdr = xdrdata.unpack_int(), xdrdata.unpack_int() # TODO: what when version 1? #TODO: what when not xdr?
+    creation =  xdrdata.unpack_string()
+    arch =  xdrdata.unpack_string()
+    os =  xdrdata.unpack_string()
+    release =  xdrdata.unpack_string()
+    text = xdrdata.unpack_string()
+
+
+    dim = xdrdata.unpack_int() # so it gives 1 (as in gdl) # TODO: I don't think will have dim>1 but need to check
+    arr_size = xdrdata.unpack_farray(dim + 2, xdrdata.unpack_int) # [1, 8, 1] = Main structure for the data
+    mainsize = arr_size[2] # number of upper level strs
+    skeleton = read_struct_skeleton(xdrdata)
+    struct_to_data(xdrdata, skeleton)
+    xdrdata.done()
+    skeleton['HEADER'] = OrderedDict([('VERSION', version), ('XDR', xdr), ('CREATION', creation),
+                                      ('IDL_VERSION', OrderedDict([('ARCH', arch),
+                                                                   ('OS', os),
+                                                                   ('RELEASE', release)])),
+                                      ('TEXT', text)])
+    # TODO: for python >= 3.2; so we can keep the original order as how it's stored in the file
```

Comment:
I don't think it really matters, even in IDL is not kept in the "same" structure. I thought since we were using `OrderedDict` it could have been better to have the `HEADER` at the beginning... but not that it matters..